Compilation error in a custom sensor using "millis()"

### The problem

Compilation error .... Compile fine with the stable release

### Which version of ESPHome has the issue?

2021.10.0

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2021.9.7

### What platform are you using?

ESP32

### Board

WROOM32

### Component causing the issue

_No response_

### Example YAML snippet

```yaml
esphome:
  name: antbms
  platform: ESP32
  board: esp32dev
  includes:
    - antbms_custom_sensor.h

wifi:
  ssid:  !secret esphome_ssid  #HOME
  password: !secret esphome_password

  # Enable fallback hotspot (captive portal) in case wifi connection fails

captive_portal:


# Enable logging
logger:

# Enable Home Assistant API
api:

ota:

switch:
  - platform: restart
    name: "ANTBMS_restart"
    
sensor:
  - platform: custom
    id: ANTBMS_id
    lambda: |-
     auto ANTBMS = new ANTBMSCustomSensor();
     App.register_component(ANTBMS);
     return {ANTBMS->ANTBMS_nbr_cell , ANTBMS->ANTBMS_soc , ANTBMS->ANTBMS_cell_av , ANTBMS->ANTBMS_cell_min ,  ANTBMS->ANTBMS_cell_max ,  ANTBMS->ANTBMS_cell_1 , ANTBMS->ANTBMS_cell_2 , ANTBMS->ANTBMS_cell_3 , ANTBMS->ANTBMS_cell_4 , ANTBMS->ANTBMS_cell_5 , ANTBMS->ANTBMS_cell_6 , ANTBMS->ANTBMS_cell_7 , ANTBMS->ANTBMS_cell_8 , ANTBMS->ANTBMS_cell_9 , ANTBMS->ANTBMS_cell_10 , ANTBMS->ANTBMS_cell_11 , ANTBMS->ANTBMS_cell_12 , ANTBMS->ANTBMS_cell_13 , ANTBMS->ANTBMS_cell_14 , ANTBMS->ANTBMS_cell_15 , ANTBMS->ANTBMS_cell_16 , ANTBMS->ANTBMS_t_bal , ANTBMS->ANTBMS_t_fet , ANTBMS->ANTBMS_t_1 , ANTBMS->ANTBMS_t_2 , ANTBMS->ANTBMS_t_3 , ANTBMS->ANTBMS_t_4 , ANTBMS->ANTBMS_puiss , ANTBMS->ANTBMS_volt , ANTBMS->ANTBMS_amp , ANTBMS->ANTBMS_Ah_install , ANTBMS->ANTBMS_Ah_rest , ANTBMS->ANTBMS_Ah_Total , ANTBMS->ANTBMS_charge_idx ,  ANTBMS->ANTBMS_discharge_idx , ANTBMS->ANTBMS_balance_idx , ANTBMS->ANTBMS_cells_balanced_idx};

    sensors:  
    - name: "ANTBMS_nbr_cell"  #1
      accuracy_decimals: 1
      icon: mdi:flash
      
    - name: "ANTBMS_soc"   #2
      unit_of_measurement: "%"
      accuracy_decimals: 1
      icon: mdi:percent  
    
    - name: "ANTBMS_cell_av"  #3
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      
    - name: "ANTBMS_cell_min" #4
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      id: ANTBMS_cell_min
      
    - name: "ANTBMS_cell_max"   #5
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      id: ANTBMS_cell_max
    
    - name: "ANTBMS_cell_1"   #6
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      
    - name: "ANTBMS_cell_2"   #7
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
  
    - name: "ANTBMS_cell_3"   #8
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      
    - name: "ANTBMS_cell_4"   #9
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      
    - name: "ANTBMS_cell_5"   #10
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      
    - name: "ANTBMS_cell_6"   #11
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium 
      
    - name: "ANTBMS_cell_7"   #12
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      
    - name: "ANTBMS_cell_8"    #13
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium  
    
    - name: "ANTBMS_cell_9"   #14
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
    
    - name: "ANTBMS_cell_10"   #15
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      
    - name: "ANTBMS_cell_11"   #16
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      
    - name: "ANTBMS_cell_12"   #17
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium  
      
    - name: "ANTBMS_cell_13"   #18
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      
    - name: "ANTBMS_cell_14"  #19
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
      
    - name: "ANTBMS_cell_15"  #19
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium
    
    - name: "ANTBMS_cell_16"  #19
      unit_of_measurement: "V"
      accuracy_decimals: 3
      icon: mdi:battery-medium  
      
    - name: "ANTBMS_t_bal"  #20
      unit_of_measurement: "C"
      accuracy_decimals: 1
      icon: mdi:thermometer  
      
    - name: "ANTBMS_t_fet" #21
      unit_of_measurement: "C"
      accuracy_decimals: 1
      icon: mdi:thermometer
      
    - name: "ANTBMS_t_1"  #22
      unit_of_measurement: "C"
      accuracy_decimals: 1
      icon: mdi:thermometer
      
    - name: "ANTBMS_t_2"  #23
      unit_of_measurement: "C"
      accuracy_decimals: 1
      icon: mdi:thermometer
      
    - name: "ANTBMS_t_3"  #23
      unit_of_measurement: "C"
      accuracy_decimals: 1
      icon: mdi:thermometer
      
    - name: "ANTBMS_t_4"  #23
      unit_of_measurement: "C"
      accuracy_decimals: 1
      icon: mdi:thermometer
      
    - name: "ANTBMS_puiss"  #24
      unit_of_measurement: "W"
      accuracy_decimals: 1
      icon: mdi:power
      
    - name: "ANTBMS_volt"  #25
      unit_of_measurement: "V"
      accuracy_decimals: 1
      icon: mdi:flash
      
    - name: "ANTBMS_amp"  #26
      unit_of_measurement: "A"
      accuracy_decimals: 1
      icon: mdi:power
    
    - name: "ANTBMS_Ah_install"  #27
      unit_of_measurement: "Ah"
      accuracy_decimals: 1
      icon: mdi:power
      
    - name: "ANTBMS_Ah_rest"  #28
      unit_of_measurement: "Ah"
      accuracy_decimals: 1
      icon: mdi:power
      
    - name: "ANTBMS_Ah_Total"  #29
      unit_of_measurement: "Ah"
      accuracy_decimals: 1
      icon: mdi:power
      
    - name: "ANTBMS_charge_idx"  #30
      id: "ANTBMS_charge_idx"
      accuracy_decimals: 0
      icon: mdi:power
      
    - name: "ANTBMS_discharge_idx"  #31
      id: "ANTBMS_discharge_idx"
      accuracy_decimals: 0
      icon: mdi:power
      
    - name: "ANTBMS_balance_idx"  #32
      id: "ANTBMS_balance_idx"
      accuracy_decimals: 0
      icon: mdi:power
      
    - name: "ANTBMS_cells_balanced_idx"  #33
      id: "ANTBMS_cells_balanced_idx"
      accuracy_decimals: 0
      icon: mdi:power
 
  - platform: template
    name: "ANTBMS_delta_maxmin"
    unit_of_measurement: "V"
    accuracy_decimals: 3
    update_interval: 5s
    icon: mdi:power
    lambda: return ( id(ANTBMS_cell_max).state - id(ANTBMS_cell_min).state);  

  
  
  - platform: wifi_signal
    name: "ANTBMS_WiFi puissance"
    update_interval: 10s     
    
  - platform: uptime
    name: "up"
    id: uptime_sec  
      
binary_sensor:
  - platform: status
    name: "ANTBMS_status"
    
    
text_sensor:
  - platform: template
    name: ANTBMS_uptime
    lambda: |-
      int seconds = (id(uptime_sec).state);
      int days = seconds / (24 * 3600);
      seconds = seconds % (24 * 3600); 
      int hours = seconds / 3600;
      seconds = seconds % 3600;
      int minutes = seconds /  60;
      seconds = seconds % 60;
      return { (String(days) +"d " + String(hours) +"h " + String(minutes) +"m "+ String(seconds) +"s").c_str() };
    icon: mdi:clock-start
    update_interval: 113s
    
  - platform: template
    name: ANTBMS_discharge
    lambda: |-
      String Discharge_St[16]={"OFF","ON","cell overdischarge","overcurrent","4","pack overdischarge",
      "bat overtemp","MOSFET overtemp","Abnormal current","battery is not detected",
      "PCB overtemp","charge MOSFET turn on","shortcircuit","Discharge MOSFET abnormality",
      "Start exception","Manual off"};
      if(id(ANTBMS_discharge_idx).has_state() )
        return {  ( Discharge_St[ (int)(id(ANTBMS_discharge_idx).state) ] ).c_str() };
      else
        return {"Not available yet"};
    icon: mdi:flash
    update_interval: 5s
    
  - platform: template
    name: ANTBMS_charge
    lambda: |-
      String Charge_St[16]={"OFF","ON","overcharge","overcurrent","batt full","pack overvoltage",
      "bat overtemp","MOSFET overtemp","abnormal current","bat not detected",
      "PCB overtemp","11-undefined","12-undefined","Discharge MOSFET abnormality","14","Manual off"};
      if(id(ANTBMS_charge_idx).has_state() )
        return { (Charge_St[(int)(id(ANTBMS_charge_idx).state)]).c_str()  };
      else
        return {"Not available yet"};
    icon: mdi:flash
    update_interval: 5s
    
    
  - platform: template
    name: ANTBMS_balance
    lambda: |-
      String BAL_St[11]={"OFF","limit trigger exceeds","Ch Vdiff high","overtemp","ACTIVE",
      "5-udef","6-udef","7-udef","8-udef","9-udef","PCB Overtemp"};
      if(id(ANTBMS_balance_idx).has_state() )
        return {(BAL_St[(int)(id(ANTBMS_balance_idx).state)]).c_str()};
      else
        return {"Not available yet"};  
    icon: mdi:flash
    update_interval: 5s
    
  - platform: template
    name: ANTBMS_cells_balanced
    lambda: |-
      if(id(ANTBMS_cells_balanced_idx).has_state() )
        return { ( String( (int)(id(ANTBMS_cells_balanced_idx).state) ) + " BIN").c_str() };
      else
        return {"Not available yet"};  
    icon: mdi:flash
    update_interval: 5s
```


### Anything in the logs that might be useful for us?

```txt
INFO Reading configuration /config/esphome/antbms.yaml...
INFO Generating C++ source...
INFO Compiling app...
INFO Running:  platformio run -d /config/esphome/antbms
Processing antbms (board: esp32dev; framework: arduino; platform: espressif32 @ 3.3.2)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
PACKAGES: 
 - framework-arduinoespressif32 3.10006.210326 (1.0.6) 
 - tool-esptoolpy 1.30100.210531 (3.1.0) 
 - toolchain-xtensa32 2.50200.97 (5.2.0)
LDF Modes: Finder ~ off, Compatibility ~ soft
Library Manager: Installing Hash
Library Manager: Already installed, built-in library
Dependency Graph
|-- <AsyncTCP-esphome> 1.2.2
|-- <WiFi> 1.0
|-- <FS> 1.0
|-- <Update> 1.0
|-- <ESPAsyncWebServer-esphome> 1.3.0
|   |-- <AsyncTCP-esphome> 1.2.2
|-- <DNSServer> 1.1.0
|-- <ESPmDNS> 1.0
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/api/api_connection.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/api/api_frame_helper.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/api/api_pb2.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/api/api_pb2_service.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/api/api_server.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/api/list_entities.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/api/proto.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/api/subscribe_state.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/api/user_services.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/api/util.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/binary_sensor/automation.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/binary_sensor/binary_sensor.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/binary_sensor/filter.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/custom/sensor/custom_sensor.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/esp32/core.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/esp32/gpio_arduino.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/esp32/gpio_idf.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/esp32/preferences.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/logger/logger.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/mdns/mdns_component.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/mdns/mdns_esp32_arduino.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/mdns/mdns_esp8266.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/mdns/mdns_esp_idf.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/network/util.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/ota/ota_backend_arduino_esp32.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/ota/ota_backend_arduino_esp8266.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/ota/ota_backend_esp_idf.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/ota/ota_component.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/restart/restart_switch.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/sensor/automation.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/sensor/filter.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/sensor/sensor.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/socket/bsd_sockets_impl.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/socket/lwip_raw_tcp_impl.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/status/status_binary_sensor.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/switch/automation.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/switch/switch.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/template/sensor/template_sensor.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/template/text_sensor/template_text_sensor.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/text_sensor/filter.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/text_sensor/text_sensor.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/uptime/uptime_sensor.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/web_server_base/web_server_base.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/wifi/wifi_component.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/wifi/wifi_component_esp32_arduino.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/wifi/wifi_component_esp_idf.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/components/wifi_signal/wifi_signal_sensor.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/core/application.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/core/color.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/core/component.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/core/controller.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/core/helpers.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/core/log.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/core/scheduler.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/esphome/core/util.cpp.o
Compiling /data/antbms/.pioenvs/antbms/src/main.cpp.o
Generating partitions /data/antbms/.pioenvs/antbms/partitions.bin
Compiling /data/antbms/.pioenvs/antbms/libeb0/AsyncTCP-esphome/AsyncTCP.cpp.o
Compiling /data/antbms/.pioenvs/antbms/lib64d/WiFi/ETH.cpp.o
Compiling /data/antbms/.pioenvs/antbms/lib64d/WiFi/WiFi.cpp.o
In file included from src/main.cpp:68:0:
src/antbms_custom_sensor.h: In member function 'virtual void ANTBMSCustomSensor::update()':
src/antbms_custom_sensor.h:76:40: error: call of overloaded 'millis()' is ambiguous
   unsigned long currentMillis = millis();
                                        ^
In file included from /data/cache/platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:35:0,
                 from /data/cache/platformio/packages/framework-arduinoespressif32/libraries/WiFi/src/WiFiUdp.h:38,
                 from /data/cache/platformio/packages/framework-arduinoespressif32/libraries/DNSServer/src/DNSServer.h:3,
                 from src/esphome/components/captive_portal/captive_portal.h:6,
                 from src/esphome.h:18,
                 from src/main.cpp:3:
/data/cache/platformio/packages/framework-arduinoespressif32/cores/esp32/esp32-hal.h:104:15: note: candidate: long unsigned int millis()
 unsigned long millis();
               ^
In file included from src/esphome/components/binary_sensor/automation.h:7:0,
                 from src/esphome.h:15,
                 from src/main.cpp:3:
src/esphome/core/hal.h:37:10: note: candidate: uint32_t esphome::millis()
 uint32_t millis();
          ^
Compiling /data/antbms/.pioenvs/antbms/lib64d/WiFi/WiFiAP.cpp.o
Archiving /data/antbms/.pioenvs/antbms/libeb0/libAsyncTCP-esphome.a
Compiling /data/antbms/.pioenvs/antbms/lib64d/WiFi/WiFiClient.cpp.o
*** [/data/antbms/.pioenvs/antbms/src/main.cpp.o] Error 1
========================= [FAILED] Took 84.52 seconds =========================
```


### Additional information

The antbms_custom_sensor.h file

```#define RXD2 17
#define TXD2 16
#define DATA_LENGTH 140

byte cmd[] = {0x5A, 0x5A, 0x00, 0x00, 0x01, 0x01};
byte startMark [] = { 0xAA, 0x55, 0xAA, 0xFF};
byte incomingByte[DATA_LENGTH] = {0};
byte temp;


int nbr_cell,soc,cell_av,cell_min,cell_max,cell_1,cell_2,cell_3,cell_4,cell_5,cell_6,cell_7,cell_8,cell_9,cell_10,cell_11,cell_12,cell_13,cell_14,cell_15,cell_16,
t_bal, t_fet, t_1, t_2, t_3, t_4, puiss;
int charge_idx=0, discharge_idx=0, balance_idx=0, cells_balanced_idx=0;
float volt,amp, Ah_install, Ah_rest, Ah_Total;


class ANTBMSCustomSensor : public PollingComponent, public Sensor 
{
public:
	Sensor *ANTBMS_nbr_cell = new Sensor();
	Sensor *ANTBMS_soc = new Sensor();
	Sensor *ANTBMS_cell_av = new Sensor();
	Sensor *ANTBMS_cell_min = new Sensor();
	Sensor *ANTBMS_cell_max = new Sensor();
	Sensor *ANTBMS_cell_1 = new Sensor();
	Sensor *ANTBMS_cell_2 = new Sensor();
	Sensor *ANTBMS_cell_3 = new Sensor();
	Sensor *ANTBMS_cell_4 = new Sensor();
	Sensor *ANTBMS_cell_5 = new Sensor();
	Sensor *ANTBMS_cell_6 = new Sensor();
	Sensor *ANTBMS_cell_7 = new Sensor();
	Sensor *ANTBMS_cell_8 = new Sensor();
	Sensor *ANTBMS_cell_9 = new Sensor();
	Sensor *ANTBMS_cell_10 = new Sensor();
	Sensor *ANTBMS_cell_11 = new Sensor();
	Sensor *ANTBMS_cell_12 = new Sensor();
    Sensor *ANTBMS_cell_13 = new Sensor();
	Sensor *ANTBMS_cell_14 = new Sensor();
	Sensor *ANTBMS_cell_15 = new Sensor();
	Sensor *ANTBMS_cell_16 = new Sensor();
	Sensor *ANTBMS_t_bal = new Sensor();
	Sensor *ANTBMS_t_fet = new Sensor();
	Sensor *ANTBMS_t_1 = new Sensor();
	Sensor *ANTBMS_t_2 = new Sensor();
	Sensor *ANTBMS_t_3 = new Sensor();
	Sensor *ANTBMS_t_4 = new Sensor();
	Sensor *ANTBMS_puiss = new Sensor();
	
	Sensor *ANTBMS_volt = new Sensor();
	Sensor *ANTBMS_amp = new Sensor();
	Sensor *ANTBMS_Ah_install = new Sensor();
	Sensor *ANTBMS_Ah_rest = new Sensor();
	Sensor *ANTBMS_Ah_Total = new Sensor();
	
	Sensor *ANTBMS_charge_idx = new Sensor();
	Sensor *ANTBMS_discharge_idx = new Sensor();
	Sensor *ANTBMS_balance_idx = new Sensor();
	Sensor *ANTBMS_cells_balanced_idx = new Sensor();

	unsigned long lastMillis=0;
    long previousMillis = 0;
    long interval = 3000;
    double temp;
	
	ANTBMSCustomSensor() : PollingComponent(3000) {}

	void setup() override
	{
		Serial2.begin(19200,SERIAL_8N1,RXD2,TXD2);	
	}
     
	void update() override 
	{
		unsigned long currentMillis = millis();
		
		if (currentMillis - previousMillis > interval) 
	    { 
					
		    Serial2.flush(); 
		    Serial2.write(cmd, sizeof(cmd)); 
			
			for (int i = 0; i < (DATA_LENGTH); i++) 
			{
			   while (!Serial2.available());
			   temp = Serial2.read();
			   incomingByte[i] = temp;
			}
		
			if ((incomingByte[0] == startMark[0]) && (incomingByte[1] == startMark [1]) && (incomingByte[2] == startMark[2]) && (incomingByte[3] == startMark[3]))

			{
			
				extract_value();
			
				ANTBMS_nbr_cell->publish_state(nbr_cell);
				ANTBMS_soc ->publish_state(soc);
				ANTBMS_cell_av ->publish_state(((float)cell_av)/1000.0);
				ANTBMS_cell_min ->publish_state(((float)cell_min/1000.0));
				ANTBMS_cell_max ->publish_state(((float)cell_max/1000.0));
				ANTBMS_cell_1 ->publish_state(((float)cell_1/1000.0));
				ANTBMS_cell_2 ->publish_state(((float)cell_2/1000.0));
				ANTBMS_cell_3 ->publish_state(((float)cell_3/1000.0));
				ANTBMS_cell_4 ->publish_state(((float)cell_4/1000.0));
				ANTBMS_cell_5 ->publish_state(((float)cell_5/1000.0));
				ANTBMS_cell_6 ->publish_state(((float)cell_6/1000.0));
				ANTBMS_cell_7 ->publish_state(((float)cell_7/1000.0));
				ANTBMS_cell_8 ->publish_state(((float)cell_8/1000.0));
				ANTBMS_cell_9 ->publish_state(((float)cell_9/1000.0));
				ANTBMS_cell_10 ->publish_state(((float)cell_10/1000.0));
				ANTBMS_cell_11 ->publish_state(((float)cell_11/1000.0));
				ANTBMS_cell_12 ->publish_state(((float)cell_12/1000.0));	
				ANTBMS_cell_13 ->publish_state(((float)cell_13/1000.0));
				ANTBMS_cell_14 ->publish_state(((float)cell_14/1000.0));
				ANTBMS_cell_15 ->publish_state(((float)cell_15/1000.0));
				ANTBMS_cell_16 ->publish_state(((float)cell_16/1000.0));
				ANTBMS_t_bal ->publish_state(t_bal);
				ANTBMS_t_fet ->publish_state(t_fet);
				ANTBMS_t_1 ->publish_state(t_1);
				ANTBMS_t_2 ->publish_state(t_2);
				ANTBMS_t_3 ->publish_state(t_3);
				ANTBMS_t_4 ->publish_state(t_4);
				ANTBMS_puiss ->publish_state(puiss);
	
				ANTBMS_volt ->publish_state((volt/10));
				ANTBMS_amp ->publish_state(-(amp/10));
				ANTBMS_Ah_install ->publish_state((Ah_install/1000000));
				ANTBMS_Ah_rest ->publish_state((Ah_rest/1000000));
				ANTBMS_Ah_Total ->publish_state((Ah_Total/1000));
				
				ANTBMS_charge_idx ->publish_state(charge_idx);
				ANTBMS_discharge_idx ->publish_state(discharge_idx);
				ANTBMS_balance_idx ->publish_state(balance_idx);
				ANTBMS_cells_balanced_idx ->publish_state(cells_balanced_idx);
			
				//print_value();
			}
			else
			{
				ESP_LOGD("customTest", "%s" , "erreur trame");
			}
		}
	}

	void extract_value()
	{
		uint32_t tmp = ((((uint8_t)incomingByte[70])<< 24) + (((uint8_t)incomingByte[71])<< 16)+ (((uint8_t)incomingByte[72])<< 8)+ ((uint8_t)incomingByte[73]));
		if (tmp > 2147483648){
			amp = (-(2*2147483648)+tmp);
		}
		else{ 
			amp = tmp;
		}
		uint32_t tmp2 = ((((uint8_t)incomingByte[111])<< 24) + (((uint8_t)incomingByte[112])<< 16)+ (((uint8_t)incomingByte[113])<< 8)+ ((uint8_t)incomingByte[114]));
		if (tmp2 > 2147483648)
		{
			puiss=-(-(2*2147483648)+tmp2);
		}
		else
		{ 
			puiss = -tmp2;
		}
		nbr_cell = (uint8_t)incomingByte[123]; 
		soc = (uint8_t)incomingByte[74]; 
		volt = ((((uint8_t)incomingByte[4])<< 8) + (uint8_t)incomingByte[5]); 
		cell_av = ((((uint8_t)incomingByte[121])<< 8) + (uint8_t)incomingByte[122]); 
		cell_min = ((((uint8_t)incomingByte[119])<< 8) + (uint8_t)incomingByte[120]); 
		cell_max = ((((uint8_t)incomingByte[116])<< 8) + (uint8_t)incomingByte[117]); 
		t_fet = ((((uint8_t)incomingByte[91])<< 8) + (uint8_t)incomingByte[92]); 
		t_bal = ((((uint8_t)incomingByte[93])<< 8) + (uint8_t)incomingByte[94]); 
		t_1 = ((((uint8_t)incomingByte[95])<< 8) + (uint8_t)incomingByte[96]); 
		t_2 = ((((uint8_t)incomingByte[97])<< 8) + (uint8_t)incomingByte[98]); 
		t_3 = ((((uint8_t)incomingByte[99])<< 8) + (uint8_t)incomingByte[100]); 
		t_4 = ((((uint8_t)incomingByte[101])<< 8) + (uint8_t)incomingByte[102]); 
		cell_1 = ((((uint8_t)incomingByte[6])<< 8) + (uint8_t)incomingByte[7]); 
		cell_2 = ((((uint8_t)incomingByte[8])<< 8) + (uint8_t)incomingByte[9]); 
		cell_3 = ((((uint8_t)incomingByte[10])<< 8) + (uint8_t)incomingByte[11]); 
		cell_4 = ((((uint8_t)incomingByte[12])<< 8) + (uint8_t)incomingByte[13]); 
		cell_5 = ((((uint8_t)incomingByte[14])<< 8) + (uint8_t)incomingByte[15]); 
		cell_6 = ((((uint8_t)incomingByte[16])<< 8) + (uint8_t)incomingByte[17]); 
		cell_7 = ((((uint8_t)incomingByte[18])<< 8) + (uint8_t)incomingByte[19]); 
		cell_8 = ((((uint8_t)incomingByte[20])<< 8) + (uint8_t)incomingByte[21]); 
		cell_9 = ((((uint8_t)incomingByte[22])<< 8) + (uint8_t)incomingByte[23]); 
		cell_10 = ((((uint8_t)incomingByte[24])<< 8) + (uint8_t)incomingByte[25]); 
		cell_11 = ((((uint8_t)incomingByte[26])<< 8) + (uint8_t)incomingByte[27]); 
		cell_12 = ((((uint8_t)incomingByte[28])<< 8) + (uint8_t)incomingByte[29]); 
		cell_13 = ((((uint8_t)incomingByte[30])<< 8) + (uint8_t)incomingByte[31]); 
		cell_14 = ((((uint8_t)incomingByte[32])<< 8) + (uint8_t)incomingByte[33]);
        cell_15 = ((((uint8_t)incomingByte[34])<< 8) + (uint8_t)incomingByte[35]);		
		cell_16 = ((((uint8_t)incomingByte[36])<< 8) + (uint8_t)incomingByte[37]);
		Ah_install = ((((uint8_t)incomingByte[75])<< 24) + (((uint8_t)incomingByte[76])<< 16)+ (((uint8_t)incomingByte[77])<< 8)+ ((uint8_t)incomingByte[78]));
		Ah_rest = ((((uint8_t)incomingByte[79])<< 24) + (((uint8_t)incomingByte[80])<< 16)+ (((uint8_t)incomingByte[81])<< 8)+ ((uint8_t)incomingByte[82]));
		Ah_Total = ((((uint8_t)incomingByte[83])<< 24) + (((uint8_t)incomingByte[84])<< 16)+ (((uint8_t)incomingByte[85])<< 8)+ ((uint8_t)incomingByte[86]));
		
		charge_idx = (uint8_t)incomingByte[103];
		discharge_idx = (uint8_t)incomingByte[104];
		balance_idx = (uint8_t)incomingByte[105];
		cells_balanced_idx = ((((uint8_t)incomingByte[132])<< 8) + (uint8_t)incomingByte[135]);
	}
};```

--- a/esphome/core/config.py
+++ b/esphome/core/config.py
@@ -225,6 +225,8 @@ async def add_arduino_global_workaround():
     # Define a hacky macro so that the call is never ambiguous
     # and always uses the esphome namespace one.
     # See also https://github.com/esphome/issues/issues/2510
+    # Priority -999 so that it runs before adding includes, as those
+    # also might reference these symbols
     for line in ARDUINO_GLUE_CODE.splitlines():
         cg.add_global(cg.RawStatement(line))
 

--- a/esphome/core/config.py
+++ b/esphome/core/config.py
@@ -206,6 +206,28 @@ def include_file(path, basename):
         cg.add_global(cg.RawStatement(f'#include "{basename}"'))
 
 
+ARDUINO_GLUE_CODE = """\
+#define yield() esphome::yield()
+#define millis() esphome::millis()
+#define delay(x) esphome::delay(x)
+#define delayMicroseconds(x) esphome::delayMicroseconds(x)
+"""
+
+
+@coroutine_with_priority(-999.0)
+async def add_arduino_global_workaround():
+    # The Arduino framework defined these itself in the global
+    # namespace. For the esphome codebase that is not a problem,
+    # but when custom code
+    #   1. writes `millis()` for example AND
+    #   2. has `using namespace esphome;` like our guides suggest
+    # Then the compiler will complain that the call is ambiguous
+    # Define a hacky macro so that the call is never ambiguous
+    # and always uses the esphome namespace one.
+    for line in ARDUINO_GLUE_CODE.splitlines():
+        cg.add_global(cg.RawStatement(line))
+
+
 @coroutine_with_priority(-1000.0)
 async def add_includes(includes):
     # Add includes at the very end, so that the included files can access global variables
@@ -286,6 +308,9 @@ async def to_code(config):
     cg.add_build_flag("-Wno-unused-variable")
     cg.add_build_flag("-Wno-unused-but-set-variable")
     cg.add_build_flag("-Wno-sign-compare")
+
+    if CORE.using_arduino:
+        CORE.add_job(add_arduino_global_workaround)
 
     if config[CONF_INCLUDES]:
         CORE.add_job(add_includes, config[CONF_INCLUDES])

--- a/esphome/core/config.py
+++ b/esphome/core/config.py
@@ -224,6 +224,7 @@ async def add_arduino_global_workaround():
     # Then the compiler will complain that the call is ambiguous
     # Define a hacky macro so that the call is never ambiguous
     # and always uses the esphome namespace one.
+    # See also https://github.com/esphome/issues/issues/2510
     for line in ARDUINO_GLUE_CODE.splitlines():
         cg.add_global(cg.RawStatement(line))
 

